### PR TITLE
Optimise WCIF endpoints

### PR DESCRIPTION
**Pull request by jonatanklosko**
Roughly 2-3x improvement for GET and ~5-8x for PATCH, as from my local tests (tested with a couple larger comps and no results).

The PATCH improvement concerns competitions with `allow_registration_without_qualification = false`, which is both competitions without qualification and competitions with qualification where enforcing is enabled.

### CODE REVIEW

File: WcaOnRails/app/models/competition.rb [hunk 1]
```diff
@@ -1637,6 +1637,7 @@ def persons_wcif(authorized: false)
       { user: {
         person: [:ranksSingle, :ranksAverage],
       } },
+      :wcif_extensions,
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:1640]: Without preloading we emitted a single additional query per registration.

File: WcaOnRails/app/models/competition.rb [hunk 2]
```diff
@@ -1734,7 +1735,7 @@ def set_wcif_events!(wcif_events, current_user)
 
     # Create missing events.
     wcif_events.each do |wcif_event|
-      event_found = competition_events.find_by_event_id(wcif_event["id"])
+      event_found = competition_events.find { |ce| ce.event_id == wcif_event["id"] }
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:1738]: We call `self.competition_events` above which loads the whole list into memory. By using `#find` we can do lookup in that list, whereas `find_by_event_id` would **always** make a query.

Same below.

File: WcaOnRails/app/models/competition.rb [hunk 3]
```diff
@@ -1783,7 +1784,9 @@ def update_persons_wcif!(wcif_persons, current_user)
       # NOTE: person doesn't necessarily have corresponding registration (e.g. registratinless organizer/delegate).
       if wcif_person["roles"]
         roles = wcif_person["roles"] - ["delegate", "trainee-delegate", "organizer"] # These three are added on the fly.
-        registration.update!(roles: roles)
+        # The additional roles are only for WCIF purposes and we don't validate them,
+        # so we can safely skip validations by using update_attribute
+        registration.update_attribute(:roles, roles)
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:1789]: Validating a registration involves a couple additional queries. We could avoid some of them via preload, but if the competition has qualifications we make a query checking against competitor's prior results, which we can't preload reasonably. In this case we only want to update the roles and can skip the validations altogether.

**Comment by coder13** [comment]
Great stuff! I'm in no place to review the improvements but they seem to make sense to me!

**Comment by jonatanklosko** [comment]
The CI fails with unrelated tests, same as on main, so I assume it's expected.

**Comment by timreyn** [comment]
Thanks so much, @jonatanklosko !!

**Review by gregorbg — APPROVED** [review]
Thanks for a couple of short and concise but important optimizations!

**Pull request merged**